src/main/scala/org/apache/spark/sql/kinesis/KinesisSourceRDD.scala

```diff
@@ -164,12 +170,27 @@ private[kinesis] class KinesisSourceRDD(
             val currentTimestamp: Long = System.currentTimeMillis
             if (canFetchMoreRecords(currentTimestamp)) {
               val shardInterator = getShardIterator()
+              // Delay if last call within a second.
+              if (lastGetRecords > 0 && (currentTimestamp - lastGetRecords) < idleTimeBetweenReads)
+              {
+                val delayMs: Long = idleTimeBetweenReads - (currentTimestamp - lastGetRecords)
+                logInfo(s"Sleeping for ${delayMs}ms")
+                Thread.sleep(delayMs)
+              }
+
               val records: GetRecordsResult = kinesisReader.getKinesisRecords(getShardIterator,
                 recordPerRequest)
               // de-aggregate records
                val deaggregateRecords = kinesisReader.deaggregateRecords(records.getRecords, null)
                fetchedRecords = deaggregateRecords.asScala.toArray
               _shardIterator = records.getNextShardIterator
+
+              // Update lastGetRecords timestamp only if got records, so we
+              // don't delay the iterations to get to the beginning of the
+              // shard
+              if (fetchedRecords.length > 0) {
+                lastGetRecords = System.currentTimeMillis
```

Comment:
If we really want to add delay, we should add it when fetchedRecords.length is 0. But then https://github.com/qubole/kinesis-sql/blob/master/src/main/scala/org/apache/spark/sql/kinesis/KinesisReader.scala#L68. will handle it if we are making too many requests